File: data_baseline/openrouter__qwen__qwen3-coder-plus/scala/rep5/Main.scala
//> using scala "2.13"  
//> using dep "org.http4s::http4s-blaze-server:0.23.15"
//> using dep "org.http4s::http4s-dsl:0.23.15"
//> using dep "org.http4s::http4s-circe:0.23.15"
//> using dep "io.circe::circe-generic:0.14.3"
//> using dep "io.circe::circe-parser:0.14.7"

import cats.effect._
import org.http4s._
import org.http4s.dsl.Http4sDsl
import org.http4s.implicits._
import org.http4s.circe._
import io.circe.generic.semiauto._
import io.circe.JsonObject
import io.circe.syntax._
import org.http4s.server.Router
import org.http4s.server.blaze.BlazeServerBuilder
import java.time.format.DateTimeFormatter
import java.time.{ZoneOffset, ZonedDateTime}
import scala.collection.mutable
import java.util.UUID

object Main extends IOApp.Simple {

  // Data Models
  case class ErrorResponse(error: String)
  case class User(id: Int, username: String)  
  case class RegisterLoginRequest(username: String, password: String)
  case class ChangePasswordRequest(oldPassword: String, newPassword: String)  
  case class CreateTodoRequest(title: String, description: String)
  case class UpdateTodoRequest(title: Option[String], description: Option[String], completed: Option[Boolean])
  case class Todo(
    id: Int,
    title: String, 
    description: String,
    completed: Boolean,
    created_at: String,
    updated_at: String
  )

  // Define implicit Circe Encoders/Decoders 
  implicit val errorResponseEncoder = deriveEncoder[ErrorResponse]
  implicit val errorResponseDecoder = deriveDecoder[ErrorResponse] 
  implicit val userEncoder = deriveEncoder[User]
  implicit val userDecoder = deriveDecoder[User]
  implicit val registerLoginRequestEncoder = deriveEncoder[RegisterLoginRequest]
  implicit val registerLoginRequestDecoder = deriveDecoder[RegisterLoginRequest]
  implicit val changePasswordRequestEncoder = deriveEncoder[ChangePasswordRequest]
  implicit val changePasswordRequestDecoder = deriveDecoder[ChangePasswordRequest]
  implicit val createTodoRequestEncoder = deriveEncoder[CreateTodoRequest]
  implicit val createTodoRequestDecoder = deriveDecoder[CreateTodoRequest]
  implicit val updateTodoRequestEncoder = deriveEncoder[UpdateTodoRequest]
  implicit val updateTodoRequestDecoder = deriveDecoder[UpdateTodoRequest]
  implicit val todoEncoder = deriveEncoder[Todo]
  implicit val todoDecoder = deriveDecoder[Todo]
  implicit val todoListEncoder = deriveEncoder[List[Todo]]

  // Storage
  val users = mutable.Map[String, (Int, String)]()
  val userSessions = mutable.Map[String, Int]()
  val userTodos = mutable.Map[Int, mutable.Map[Int, Todo]]()

  // ID Generators
  object GenId {
    private var userCount = 0
    private var todoCount = 0
    
    def nextUser(): Int = synchronized { userCount += 1; userCount }
    def nextTodo(): Int = synchronized { todoCount += 1; todoCount }
  }

  // Utilities
  def checkValidUsername(username: String): Boolean = 
    username.length >= 3 && username.length <= 50 && username.matches("^[a-zA-Z0-9_]+$")

  def getTimestamp: String = 
    ZonedDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'"))

  def createSession: String = UUID.randomUUID().toString

  def identifyUser(request: Request[IO]): Option[Int] = 
    request.cookies.find(_.name == "session_id").map(_.content).flatMap(userSessions.get)

  def mainRoutes: HttpRoutes[IO] = {
    val dsl = Http4sDsl[IO]
    import dsl._
    
    import org.http4s.circe.CirceEntityEncoder._

    HttpRoutes.of[IO] {
      
      // Public Endpoints
      case request @ POST -> Root / "register" =>
        for {
          regPayload <- request.as[RegisterLoginRequest]
          result <- {
            if (!checkValidUsername(regPayload.username)) {
              BadRequest(ErrorResponse("Invalid username"))
            } else if (regPayload.password.length < 8) {
              BadRequest(ErrorResponse("Password too short")) 
            } else if (users.contains(regPayload.username)) {
              Conflict(ErrorResponse("Username already exists"))
            } else {
              val newUserId = GenId.nextUser()
              users.update(regPayload.username, (newUserId, regPayload.password))
              userTodos.update(newUserId, mutable.Map[Int, Todo]())
              Created(User(newUserId, regPayload.username))
            }
          }
        } yield result
      
      case request @ POST -> Root / "login" =>
        for {
          loginPayload <- request.as[RegisterLoginRequest]
          userExists = users.get(loginPayload.username).exists(_._2 == loginPayload.password)
          result <- {
            if (userExists) {
              val userId = users(loginPayload.username)._1
              val sessionId = createSession
              userSessions.update(sessionId, userId)
              
              Ok(User(userId, loginPayload.username)).map(_.addCookie(
                ResponseCookie(
                  name = "session_id",
                  content = sessionId,
                  path = Some("/"),
                  httpOnly = true
                )
              ))
            } else {
              Unauthorized(ErrorResponse("Invalid credentials"))
            }
          }
        } yield result
      
      // Private Endpoints
      case request @ POST -> Root / "logout" =>
        identifyUser(request) match {
          case Some(userId) =>
            userSessions.filterInPlace { case (_, id) => id != userId }
            Ok("")
          case None =>
            Unauthorized(ErrorResponse("Authentication required"))
        }
      
      case request @ GET -> Root / "me" =>  
        identifyUser(request) match {
          case Some(userId) =>
            users.find(_._2._1 == userId) match {
              case Some((username, (id, _))) => Ok(User(id, username))
              case None => Unauthorized(ErrorResponse("Authentication required"))
            }
          case None => 
            Unauthorized(ErrorResponse("Authentication required"))
        }
      
      case request @ PUT -> Root / "password" =>
        identifyUser(request) match {
          case Some(userId) =>
            for {
              pwdPayload <- request.as[ChangePasswordRequest]
              matchingUser = users.find(u => u._2._1 == userId && u._2._2 == pwdPayload.oldPassword)
              result <- { 
                if (matchingUser.isEmpty) {
                  Unauthorized(ErrorResponse("Invalid credentials"))
                } else if (pwdPayload.newPassword.length < 8) {
                  BadRequest(ErrorResponse("Password too short"))
                } else {
                  val userKey = matchingUser.head._1
                  users.update(userKey, (userId, pwdPayload.newPassword))
                  Ok("")
                }
              }
            } yield result
          case None =>
            Unauthorized(ErrorResponse("Authentication required"))
        }
      
      case request @ GET -> Root / "todos" =>
        identifyUser(request) match {
          case Some(userId) =>
            val userTodoList = userTodos.get(userId).map(_.values.toList.sortBy(_.id)).getOrElse(Nil)
            Ok(userTodoList)
          case None =>
            Unauthorized(ErrorResponse("Authentication required"))
        }
      
      case request @ POST -> Root / "todos" =>
        identifyUser(request) match {
          case Some(userId) =>
            for {
              todoPayload <- request.as[CreateTodoRequest]
              result <- {
                if (todoPayload.title.trim.isEmpty) {
                  BadRequest(ErrorResponse("Title is required"))
                } else {
                  val newTodoId = GenId.nextTodo()
                  val now = getTimestamp
                  val newTodo = Todo(
                    id = newTodoId,
                    title = todoPayload.title,
                    description = Option(todoPayload.description).getOrElse(""),  
                    completed = false,
                    created_at = now,
                    updated_at = now
                  )
                  
                  val userTodoMap = userTodos.getOrElseUpdate(userId, mutable.Map[Int, Todo]())
                  userTodoMap.update(newTodoId, newTodo)
                  Created(newTodo)  
                }
              }
            } yield result
          case None =>
            Unauthorized(ErrorResponse("Authentication required"))
        }
      
      case GET -> Root / "todos" / IntVar(todoId) =>
        identifyUser(request) match {
          case Some(userId) =>
            userTodos.get(userId).flatMap(_.get(todoId)) match {
              case Some(todo) => Ok(todo)
              case None => NotFound(ErrorResponse("Todo not found"))
            }
          case None =>
            Unauthorized(ErrorResponse("Authentication required"))  
        }
      
      case request @ PUT -> Root / "todos" / IntVar(todoId) =>
        identifyUser(request) match {
          case Some(userId) =>
            userTodos.get(userId).flatMap(_.get(todoId)) match {
              case Some(existingTodo) =>
                for {
                  updatePayload <- request.as[UpdateTodoRequest]
                  processedResult <- {
                    if (updatePayload.title.exists(_.trim.isEmpty)) {
                      IO.pure(BadRequest(ErrorResponse("Title is required")))
                    } else {
                      val updatedTodo = existingTodo.copy(
                        title = updatePayload.title.getOrElse(existingTodo.title),
                        description = updatePayload.description.getOrElse(existingTodo.description), 
                        completed = updatePayload.completed.getOrElse(existingTodo.completed),
                        updated_at = getTimestamp
                      )
                      
                      userTodos(userId).update(todoId, updatedTodo)
                      IO.pure(Ok(updatedTodo))
                    }
                  }
                } yield processedResult
              case None => 
                IO.pure(NotFound(ErrorResponse("Todo not found")))
            }
          case None =>
            IO.pure(Unauthorized(ErrorResponse("Authentication required")))
        }
      
      case DELETE -> Root / "todos" / IntVar(todoId) =>
        identifyUser(request) match {
          case Some(userId) =>
            userTodos.get(userId) match {
              case Some(todoMap) if todoMap.contains(todoId) =>
                todoMap.remove(todoId)
                NoContent()
              case _ =>
                NotFound(ErrorResponse("Todo not found"))
            }
          case None =>
            Unauthorized(ErrorResponse("Authentication required"))
        }
    }
  }

  def run: IO[Unit] = {
    val portOpt = sys.props.get("HTTP_PORT").flatMap(_.toIntOption).orElse(Some(8080))
    
    BlazeServerBuilder[IO]
      .bindHttp(portOpt.get, "0.0.0.0")
      .withHttpApp(Router("/" -> mainRoutes.orNotFound))
      .serve
      .compile  
      .drain
  }
}

File: data_baseline/openrouter__qwen__qwen3-coder-plus/rust/rep5/src/main.rs
use axum::{
    extract::{FromRequestParts, Path, State},
    http::{header::COOKIE, HeaderMap, StatusCode, HeaderValue},
    response::{IntoResponse, Response},
    routing::{get, post, put, delete},
    Json, Router, 
};
use chrono::{DateTime, Utc};
use lazy_static::lazy_static;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use tokio::signal;
use uuid::Uuid;

// Data Structures
#[derive(Debug, Clone, Serialize)]
pub struct User {
    pub id: i32,
    pub username: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Todo {
    pub id: i32,
    pub title: String,
    pub description: String,
    pub completed: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

// Authentication extractor
#[derive(Debug, Clone)]
pub struct AuthenticatedUser {
    pub user_id: i32,
    pub username: String,
}

#[async_trait::async_trait]
impl<S> FromRequestParts<S> for AuthenticatedUser
where
    S: Send + Sync,
{
    type Rejection = (StatusCode, Json<serde_json::Value>);

    async fn from_request_parts(parts: &mut axum::http::request::Parts, _state: &S) -> Result<Self, Self::Rejection> {
        let headers = &parts.headers;
        
        if let Some(session_id) = extract_session_id(headers) {
            // Get state from extensions or look for a better way to access it
            // Since we're inside the handler chain, we get the state differently
            // Let's implement differently by checking state availability directly
            
            // For the middleware pattern, we would need to set up state properly in extensions, 
            // but axum makes this tricky. So instead, let's modify it so the function receives state in each route
            // Instead of having a true extractor, I'll implement direct verification in protected endpoints
            
            // Just return placeholder error to indicate the issue, actually we need to implement a proper solution
            return Err((StatusCode::UNAUTHORIZED, Json(serde_json::json!({"error": "Authentication required"}))));
        }
        
        Err((StatusCode::UNAUTHORIZED, Json(serde_json::json!({"error": "Authentication required"}))))
    }
}

#[derive(Deserialize)]
struct RegisterRequest {
    username: String,
    password: String,
}

#[derive(Deserialize)]
struct LoginRequest {
    username: String,
    password: String,
}

#[derive(Deserialize)]
struct UpdatePasswordRequest {
    old_password: String,
    new_password: String,
}

#[derive(Deserialize)]
struct CreateTodoRequest {
    title: String,
    description: Option<String>,
}

#[derive(Deserialize)]
struct UpdateTodoRequest {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

#[derive(Debug, Clone)]
struct Session {
    user_id: i32,
}

#[derive(Clone)]
struct AppState {
    users: Arc<RwLock<HashMap<i32, User>>>,
    passwords: Arc<RwLock<HashMap<i32, String>>>, // Simple implementation, real apps should hash
    todos: Arc<RwLock<HashMap<i32, HashMap<i32, Todo>>>>, // user_id -> {todo_id -> todo}
    sessions: Arc<RwLock<HashMap<String, Session>>>,
    next_user_id: Arc<RwLock<i32>>,
    next_todo_id: Arc<RwLock<i32>>,
}

impl AppState {
    fn new() -> Self {
        Self {
            users: Arc::new(RwLock::new(HashMap::new())),
            passwords: Arc::new(RwLock::new(HashMap::new())),
            todos: Arc::new(RwLock::new(HashMap::new())),
            sessions: Arc::new(RwLock::new(HashMap::new())),
            next_user_id: Arc::new(RwLock::new(1)),
            next_todo_id: Arc::new(RwLock::new(1)),
        }
    }

    fn validate_username(username: &str) -> bool {
        lazy_static! {
            static ref USERNAME_REGEX: Regex = Regex::new(r"^[a-zA-Z0-9_]{3,50}$").unwrap();
        }
        USERNAME_REGEX.is_match(username)
    }

    async fn register_user(&self, req: RegisterRequest) -> Result<User, (StatusCode, String)> {
        if !Self::validate_username(&req.username) {
            return Err((StatusCode::BAD_REQUEST, "Invalid username".to_string()));
        }
        
        if req.password.len() < 8 {
            return Err((StatusCode::BAD_REQUEST, "Password too short".to_string()));
        }

        let mut users = self.users.write().unwrap();
        for user in users.values() {
            if user.username == req.username {
                return Err((StatusCode::CONFLICT, "Username already exists".to_string()));
            }
        }

        let user_id = *self.next_user_id.write().unwrap();
        *self.next_user_id.write().unwrap() += 1;

        let user = User {
            id: user_id,
            username: req.username.clone(),
        };

        users.insert(user_id, user.clone());
        drop(users);
        
        let mut passwords = self.passwords.write().unwrap();
        passwords.insert(user_id, req.password);

        Ok(user)
    }

    async fn authenticate_user(&self, username: &str, password: &str) -> Option<i32> {
        let users = self.users.read().unwrap();
        let mut target_user_id = None;
        for (user_id, user) in users.iter() {
            if user.username == username {
                target_user_id = Some(*user_id);
                break;
            }
        }

        match target_user_id {
            Some(user_id) => {
                let passwords = self.passwords.read().unwrap();
                if let Some(stored_password) = passwords.get(&user_id) {
                    if *stored_password == password {
                        Some(user_id)
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
            None => None,
        }
    }

    async fn create_session(&self, user_id: i32) -> String {
        let session_id = Uuid::new_v4().to_string();
        let session = Session { user_id };
        
        let mut sessions = self.sessions.write().unwrap();
        sessions.insert(session_id.clone(), session);
        
        session_id
    }

    async fn get_user_from_session(&self, session_id: &str) -> Option<(i32, User)> {
        let sessions = self.sessions.read().unwrap();
        if let Some(session) = sessions.get(session_id) {
            let users = self.users.read().unwrap();
            if let Some(user) = users.get(&session.user_id).cloned() {
                Some((session.user_id, user))
            } else {
                None
            }
        } else {
            None
        }
    }

    async fn invalidate_session(&self, session_id: &str) -> bool {
        let mut sessions = self.sessions.write().unwrap();
        sessions.remove(session_id).is_some()
    }

    async fn change_password(&self, user_id: i32, old_password: &str, new_password: &str) -> Result<(), (StatusCode, String)> {
        let passwords = self.passwords.read().unwrap();
        if let Some(current_password) = passwords.get(&user_id) {
            if *current_password != old_password {
                return Err((StatusCode::UNAUTHORIZED, "Invalid credentials".to_string()));
            }
        } else {
            return Err((StatusCode::UNAUTHORIZED, "Invalid credentials".to_string()));
        }
        
        if new_password.len() < 8 {
            return Err((StatusCode::BAD_REQUEST, "Password too short".to_string()));
        }

        drop(passwords);
        let mut passwords = self.passwords.write().unwrap();
        passwords.insert(user_id, new_password.to_string());
        Ok(())
    }

    async fn create_todo(&self, user_id: i32, req: CreateTodoRequest) -> Result<Todo, (StatusCode, String)> {
        if req.title.is_empty() {
            return Err((StatusCode::BAD_REQUEST, "Title is required".to_string()));
        }

        let todo_id = *self.next_todo_id.write().unwrap();
        *self.next_todo_id.write().unwrap() += 1;

        let now = Utc::now();
        let todo = Todo {
            id: todo_id,
            title: req.title,
            description: req.description.unwrap_or_default(),
            completed: false,
            created_at: now,
            updated_at: now,
        };

        let mut todos = self.todos.write().unwrap();
        let user_todos = todos.entry(user_id).or_insert_with(HashMap::new);
        user_todos.insert(todo_id, todo.clone());

        Ok(todo)
    }

    async fn get_todos_for_user(&self, user_id: i32) -> Vec<Todo> {
        let todos = self.todos.read().unwrap();
        if let Some(user_todos) = todos.get(&user_id) {
            let mut result: Vec<Todo> = user_todos.values().cloned().collect();
            result.sort_by_key(|t| t.id); // Sort by ID ascending
            result
        } else {
            vec![]
        }
    }

    async fn get_todo_for_user(&self, user_id: i32, todo_id: i32) -> Option<Todo> {
        let todos = self.todos.read().unwrap();
        if let Some(user_todos) = todos.get(&user_id) {
            user_todos.get(&todo_id).cloned()
        } else {
            None
        }
    }

    async fn update_todo(&self, user_id: i32, todo_id: i32, req: UpdateTodoRequest) -> Result<Todo, (StatusCode, String)> {
        let mut todos = self.todos.write().unwrap();
        let user_todos = todos.get_mut(&user_id);
        if user_todos.is_none() {
            return Err((StatusCode::NOT_FOUND, "Todo not found".to_string()));
        }
        let user_todos = user_todos.unwrap();

        let todo = user_todos.get_mut(&todo_id);
        if todo.is_none() {
            return Err((StatusCode::NOT_FOUND, "Todo not found".to_string()));
        }
        let todo = todo.unwrap();

        // Validate title if provided
        if let Some(ref title) = req.title {
            if title.is_empty() {
                return Err((StatusCode::BAD_REQUEST, "Title is required".to_string()));
            }
            todo.title = title.clone();
        }

        // Update other fields if provided
        if let Some(description) = req.description {
            todo.description = description;
        }
        if let Some(completed) = req.completed {
            todo.completed = completed;
        }

        todo.updated_at = Utc::now();

        Ok(todo.clone())
    }

    async fn delete_todo(&self, user_id: i32, todo_id: i32) -> bool {
        let mut todos = self.todos.write().unwrap();
        if let Some(user_todos) = todos.get_mut(&user_id) {
            user_todos.remove(&todo_id).is_some()
        } else {
            false
        }
    }
}

// Helper function to extract session ID from cookies
fn extract_session_id(headers: &HeaderMap) -> Option<String> {
    if let Some(cookie_header) = headers.get(COOKIE) {
        if let Ok(cookie_str) = cookie_header.to_str() {
            for cookie in cookie_str.split(';') {
                let cookie = cookie.trim();
                if cookie.starts_with("session_id=") {
                    return Some(cookie[11..].to_string()); // Remove "session_id=" prefix
                }
            }
        }
    }
    None
}

async fn check_authentication(headers: &HeaderMap, state: &AppState) -> Option<(i32, User)> {
    if let Some(session_id) = extract_session_id(headers) {
        return state.get_user_from_session(&session_id).await;
    }
    None
}

// Error response helper
fn error_response(status: StatusCode, message: &str) -> Response {
    (
        status,
        Json(serde_json::json!({"error": message}))
    ).into_response()
}

// Route handlers
async fn register(
    State(state): State<AppState>,
    Json(req): Json<RegisterRequest>,
) -> Result<impl IntoResponse, Response> {
    match state.register_user(req).await {
        Ok(user) => Ok((StatusCode::CREATED, Json(user))),
        Err((status, error_msg)) => Err(error_response(status, &error_msg)),
    }
}

async fn login(
    State(state): State<AppState>,
    Json(req): Json<LoginRequest>,
) -> Result<Response, Response> {
    match state.authenticate_user(&req.username, &req.password).await {
        Some(user_id) => {
            let session_id = state.create_session(user_id).await;
            
            let users = state.users.read().unwrap();
            let user = users.get(&user_id).unwrap().clone();
            drop(users);

            // Create response with custom headers
            let mut resp = Response::builder()
                .status(StatusCode::OK)
                .header("Content-Type", "application/json");

            let body = serde_json::to_vec(&user).unwrap();
            
            let res = resp
                .body(axum::body::Body::from(body))
                .map_err(|_| error_response(StatusCode::INTERNAL_SERVER_ERROR, "Failed to create response"))?;

            // We need to set the cookie in the actual response's headers
            let mut result_resp = res;
            if let Ok(set_cookie_value) = HeaderValue::from_str(&format!(
                "session_id={}; Path=/; HttpOnly",
                session_id
            )) {
                result_resp.headers_mut().insert(axum::http::header::SET_COOKIE, set_cookie_value);
            }
            
            Ok(result_resp)
        },
        None => Err(error_response(StatusCode::UNAUTHORIZED, "Invalid credentials"))
    }
}

async fn logout(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<impl IntoResponse, Response> {
    if let Some(session_id) = extract_session_id(&headers) {
        state.invalidate_session(&session_id).await;
    }
    
    Ok((StatusCode::OK, Json(serde_json::json!({}))))
}

async fn get_me(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> Result<impl IntoResponse, Response> {
    if let Some((user_id, user)) = check_authentication(&headers, &state).await {
        Ok(Json(user))
    } else {
        Err(error_response(StatusCode::UNAUTHORIZED, "Authentication required"))
    }
}

async fn update_password(
    headers: HeaderMap,
    State(state): State<AppState>,
    Json(req): Json<UpdatePasswordRequest>,
) -> Result<impl IntoResponse, Response> {
    if let Some((user_id, _)) = check_authentication(&headers, &state).await {
        match state.change_password(user_id, &req.old_password, &req.new_password).await {
            Ok(()) => Ok((StatusCode::OK, Json(serde_json::json!({})), )),
            Err((status, error_msg)) => Err(error_response(status, &error_msg)), 
        }
    } else {
        Err(error_response(StatusCode::UNAUTHORIZED, "Authentication required"))
    }
}

async fn get_todos(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> Result<impl IntoResponse, Response> {
    if let Some((user_id, _)) = check_authentication(&headers, &state).await {
        let todos = state.get_todos_for_user(user_id).await;
        Ok(Json(todos))
    } else {
        Err(error_response(StatusCode::UNAUTHORIZED, "Authentication required"))
    }
}

async fn create_todo(
    headers: HeaderMap,
    State(state): State<AppState>,
    Json(req): Json<CreateTodoRequest>,
) -> Result<impl IntoResponse, Response> {
    if let Some((user_id, _)) = check_authentication(&headers, &state).await {
        match state.create_todo(user_id, req).await {
            Ok(todo) => Ok((StatusCode::CREATED, Json(todo))),
            Err((status, error_msg)) => Err(error_response(status, &error_msg)), 
        }
    } else {
        Err(error_response(StatusCode::UNAUTHORIZED, "Authentication required"))
    }
}

async fn get_todo_by_id(
    headers: HeaderMap,
    State(state): State<AppState>,
    Path(todo_id): Path<i32>,
) -> Result<impl IntoResponse, Response> {
    if let Some((user_id, _)) = check_authentication(&headers, &state).await {
        if let Some(todo) = state.get_todo_for_user(user_id, todo_id).await {
            Ok(Json(todo))
        } else {
            Err(error_response(StatusCode::NOT_FOUND, "Todo not found"))
        }
    } else {
        Err(error_response(StatusCode::UNAUTHORIZED, "Authentication required"))
    }
}

async fn update_todo_by_id(
    headers: HeaderMap,
    State(state): State<AppState>,
    Path(todo_id): Path<i32>,
    Json(req): Json<UpdateTodoRequest>,
) -> Result<impl IntoResponse, Response> {
    if let Some((user_id, _)) = check_authentication(&headers, &state).await {
        match state.update_todo(user_id, todo_id, req).await {
            Ok(updated_todo) => Ok(Json(updated_todo)),
            Err((status, error_msg)) => Err(error_response(status, &error_msg)),
        }
    } else {
        Err(error_response(StatusCode::UNAUTHORIZED, "Authentication required"))
    }
}

async fn delete_todo_by_id(
    headers: HeaderMap,
    State(state): State<AppState>,
    Path(todo_id): Path<i32>,
) -> Result<impl IntoResponse, Response> {
    if let Some((user_id, _)) = check_authentication(&headers, &state).await {
        if state.delete_todo(user_id, todo_id).await {
            Ok(StatusCode::NO_CONTENT.into_response())
        } else {
            Err(error_response(StatusCode::NOT_FOUND, "Todo not found"))
        }
    } else {
        Err(error_response(StatusCode::UNAUTHORIZED, "Authentication required"))
    }
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut port = 8080;
    
    for i in 0..args.len() {
        if args[i] == "--port" && i + 1 < args.len() {
            port = args[i + 1].parse().expect("--port argument must be a valid number");
        }
    }
    
    println!("Starting server on port {}", port);

    let state = AppState::new();
    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(get_me))
        .route("/password", put(update_password))
        .route("/todos", get(get_todos).post(create_todo))
        .route("/todos/:id", get(get_todo_by_id).put(update_todo_by_id).delete(delete_todo_by_id))
        .with_state(state);

    let listener = tokio::net::TcpListener::bind(format!("0.0.0.0:{}", port))
        .await
        .unwrap();
        
    axum::serve(listener, app)
        .with_graceful_shutdown(shutdown_signal())
        .await
        .unwrap();
}

async fn shutdown_signal() {
    let ctrl_c = async {
        signal::ctrl_c()
            .await
            .expect("Failed to install Ctrl+C handler");
    };

    #[cfg(unix)]
    let terminate = async {
        signal::unix::signal(signal::unix::SignalKind::terminate())
            .expect("Failed to install signal handler")
            .recv()
            .await;
    };

    #[cfg(not(unix))]
    let terminate = std::future::pending::<()>();

    tokio::select! {
        _ = ctrl_c => {},
        _ = terminate => {},
    }
    
    println!("Shutting down gracefully...");
}